### ISSUE TITLE
unstack() treats string coords as objects

### ISSUE THREAD

**Issue by crusaderky** [pydata/xarray#907]
unstack() should be smart enough to recognise that all labels in a coord are strings, and convert them to numpy strings.
This is particularly relevant e.g. if you want to dump the xarray to netcdf and then read it with a non-python library.

``` python
import xarray

a = xarray.DataArray([[1,2],[3,4]], dims=['x', 'y'], coords={'x': ['x1', 'x2'], 'y': ['y1', 'y2']})
a
```

```
<xarray.DataArray (x: 2, y: 2)>
array([[1, 2],
       [3, 4]])
Coordinates:
  * y        (y) <U2 'y1' 'y2'
  * x        (x) <U2 'x1' 'x2'
```

``` python
a.stack(s=['x', 'y']).unstack('s')
```

```
<xarray.DataArray (x: 2, y: 2)>
array([[1, 2],
       [3, 4]])
Coordinates:
  * x        (x) object 'x1' 'x2'
  * y        (y) object 'y1' 'y2'
```

**Comment by shoyer** [comment]
This is a tricky one to fix. Stack is not the only operation that can convert string arrays to dtype=object -- reindex and align also do this.

The challenge is that NumPy does not support variable width strings or strings with missing values. Hence, we end up with dtype=object arrays.

In this particular case, we could preserve the string dtype by keeping track of the original dtype of the arrays as metadata, similar to how we currently use [`PandasIndexAdapter`](https://github.com/pydata/xarray/blob/884b247742af2480e269bb54c9102f08c867fdaa/xarray/core/indexing.py#L425) to preserve the original dtype on arrays converted into non-multi indexes.

**Comment by crusaderky** [comment]
Crude workaround:

```python
# Convert object coords to string
for dim in array.dims:
    if array.coords[dim].dtype == numpy.dtype('O'):
        array.coords[dim] = array.coords[dim].values.tolist()

return array
```
The above works, but if you have more than a handful elements in the coord the performance hit will be severe.

**Comment by shoyer** [comment]
Does `array.coords[dim] = array.coords[dim].values.astype(str)` work? In theory, I think it should...

**Comment by crusaderky** [comment]
Yes that works as well

**Comment by crusaderky** [comment]
This issue is still valid as of xarray 0.11.0

**Comment by benbovy** [comment]
This is fixed in v2022.6.0. Xarray's `PandasMultiIndex` wrapper keeps track of the level coordinate dtypes.
